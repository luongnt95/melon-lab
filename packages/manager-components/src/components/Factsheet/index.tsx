import React, { StatelessComponent } from 'react';
import Button from '~/blocks/Button';
import Icon from '~/blocks/Icon';
import Loading from '~/blocks/Loading';
import Spinner from '~/blocks/Spinner';

import styles from './styles.css';

export interface FactsheetProps {
  gav?: string;
  creationDate?: string;
  dataValid?: boolean;
  isCompetition?: boolean;
  loading?: boolean;
  managementReward?: string;
  name?: string;
  numberOfFunds?: string;
  performanceReward?: string;
  personalStake?: string;
  quoteAsset?: string;
  rank?: string;
  reportUrl?: string;
  scrollTo?: (id) => void;
  sharePrice?: string;
  shutdown: () => void;
  totalSupply?: string;
  tweetHref?: string;
  address?: string;
  track?: string;
  owner?: boolean;
}

const Factsheet: StatelessComponent<FactsheetProps> = ({
  gav,
  creationDate,
  dataValid,
  isCompetition,
  loading,
  managementReward,
  name,
  numberOfFunds,
  performanceReward,
  personalStake,
  quoteAsset,
  rank,
  reportUrl,
  scrollTo,
  sharePrice,
  shutdown,
  totalSupply,
  address,
  track,
  owner,
}) => {
  const scrolltoHoldings = () => scrollTo && scrollTo('holdings');

  const buildTwitterUrl = () => {
    const text = owner
      ? track !== 'live'
        ? `My #MelonFund "${name}" has a share price currently of ${sharePrice}. Have a look:`
        : `Check out my on-chain decentralized hedge fund "${name}". ` +
          `It currently has a share price of ${sharePrice}. Have a look:`
      : track !== 'live'
        ? `The #MelonFund "${name}" has a share price currently of ${sharePrice}. Have a look:`
        : `Check out this on-chain decentralized hedge fund "${name}". ` +
          `It currently has a share price of ${sharePrice}. Have a look:`;

    const url =
      track === 'live'
        ? `https://olympiad.melon.fund/#${address}`
        : `https://melon.fund/#${address}`;
    const hashtags = 'TechnologyRegulatedFunds,Melon,MelonFund';
    const via = 'melonport';
    const related = 'melonport';

    return `https://twitter.com/intent/tweet?text=${encodeURIComponent(
      text,
    )}&url=${encodeURIComponent(url)}&hashtags=${encodeURIComponent(
      hashtags,
    )}&via=${encodeURIComponent(via)}&related=${encodeURIComponent(related)}`;
  };

  return (
    <div className="factsheet">
      <style jsx>{styles}</style>
      <h3 className="factsheet__title">
        {name}
        <a
          className="factsheet__title-link"
          href={buildTwitterUrl()}
          target="_blank"
          rel="noopener noreferrer"
        >
          <Icon width="12px" height="12px" name="icons_twitter" />
        </a>
      </h3>
      {loading ? (
        <Spinner icon size="small" />
      ) : (
        <div>
          Creation date:{' '}
          <Loading loading={creationDate === '...'}>{creationDate}</Loading>
          <br />
          <Button onClick={scrolltoHoldings} style="clear">
            AUM:{' '}
            <Loading dataAvailable={dataValid} loading={gav === '...'}>
              {gav.toFixed(4)}
            </Loading>{' '}
            {quoteAsset}
          </Button>
          <Button onClick={scrolltoHoldings} style="clear">
            Share price:{' '}
            <Loading dataAvailable={dataValid} loading={sharePrice === '...'}>
              {sharePrice.toFixed(4)}
            </Loading>{' '}
            {quoteAsset}
            /Share
          </Button>
          <a href="#/ranking">
            Ranking: <Loading loading={rank === '...'}>{rank}</Loading> out of{' '}
            <Loading loading={numberOfFunds === '...'}>{numberOfFunds}</Loading>
          </a>
          <br />
          Total number of shares:{' '}
          <Loading loading={totalSupply === '...'}>{totalSupply}</Loading>
          <br />
          Shares owned by me:{' '}
          <Loading loading={personalStake === '...'}>{personalStake}</Loading>
          <hr />
          Management Reward:{' '}
          <Loading loading={managementReward === '...'}>
            {managementReward}
          </Loading>
          %<br />
          Performance Reward:{' '}
          <Loading loading={performanceReward === '...'}>
            {performanceReward}
          </Loading>
          %<hr />
          <a
            href="https://ipfs.io/ipfs/Qmc9JRw4zarrs6gJwu6tC58UAgeEujNg9VMWcH8MUEd5TW/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Contact Investors/Managers
          </a>
          <hr />
          <a href={reportUrl} target="_blank" rel="noopener noreferrer">
            Generate fund report
          </a>
          <hr />
          {!isCompetition ? (
            <Button onClick={shutdown} style="clear">
              Irreversibly shut down fund
            </Button>
          ) : (
            <div />
          )}
        </div>
      )}
    </div>
  );
};

export default Factsheet;
