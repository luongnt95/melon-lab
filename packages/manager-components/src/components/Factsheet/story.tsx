import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';
import React from 'react';
import Factsheet from './index';

const data = {
  gav: '0.9824',
  creationDate: '24. Jul 2018 09:19',
  dataValid: false,
  expectedPrize: '0.54321',
  isCompetition: false,
  loading: false,
  managementReward: '0.0000',
  name: 'Melon Fund',
  numberOfFunds: '12',
  performanceReward: '0.0000',
  personalStake: '1.0000',
  quoteAsset: 'WETH-T',
  rank: '1',
  sharePrice: '0.9824',
  shutdown: action('shitdown'),
  totalSupply: '1.0000',
  tweetHref: 'melonport',
  reportUrl: 'https://melon-reporting.now.sh/report/',
};

storiesOf('Components|Factsheet', module).add('Default', () => {
  return <Factsheet {...data} />;
});
