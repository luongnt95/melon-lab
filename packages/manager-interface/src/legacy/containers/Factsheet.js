import moment from 'moment';
import { connect } from 'react-redux';
import { compose } from 'recompose';
import Factsheet from '@melonproject/manager-components/components/Factsheet';
import { actions as appActions } from '../actions/app';
import { actions as administrationActions } from '../actions/administration';
import { networks } from '@melonproject/melon.js';

const mapStateToProps = (state, props) => ({
  creationDate:
    props.inception && moment(props.inception).format('D. MMM YYYY HH:mm'),
  personalStake: state.fund.personalStake,
  dataValid: state.ethereum.isDataValid,
  numberOfFunds: state.fund.numberOfFunds,
  quoteAsset: state.app.assetPair.quote,
  isCompetition: state.app.isCompetition,
  track: state.app.track,
  reportUrl: `https://${
    state.ethereum.network === networks.KOVAN ? 'melon' : 'olympiad'
  }-reporting.now.sh/report/${props.address}`,
});

const mapDispatchToProps = dispatch => ({
  scrollTo: target => dispatch(appActions.scrollTo(target)),
  shutdown: () => dispatch(administrationActions.shutdown()),
});

const withState = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(withState)(Factsheet);
