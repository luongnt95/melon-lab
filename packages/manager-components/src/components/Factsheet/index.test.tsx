import React from 'react';
import Factsheet from './index';

const mockCallbackShutdown = jest.fn();
const mockCallbackScrollTo = jest.fn();
const data = {
  gav: 0.9824,
  creationDate: '24. Jul 2018 09:19',
  dataValid: false,
  isCompetition: false,
  loading: false,
  managementReward: 0.0000,
  name: 'Melon Fund',
  numberOfFunds: '12',
  performanceReward: 0.0000,
  personalStake: 1.0000,
  quoteAsset: 'WETH-T',
  rank: '1',
  sharePrice: 0.9824,
  shutdown: mockCallbackShutdown,
  scrollTo: mockCallbackScrollTo,
  totalSupply: '1.0000',
  tweetHref: 'melonport',
  reportUrl: 'https://melon-reporting.now.sh/report/',
};

describe('Factsheet', () => {
  const defaultElement = <Factsheet {...data} />;
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(defaultElement);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with loading', () => {
    wrapper.setProps({ loading: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with isCompetition', () => {
    wrapper.setProps({ isCompetition: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('scrollTo event', () => {
    wrapper
      .find('Button')
      .first()
      .simulate('click');
    expect(mockCallbackScrollTo.mock.calls.length).toBe(1);
    expect(mockCallbackScrollTo.mock.calls[0][0]).toEqual('holdings');
  });

  it('shutdown event', () => {
    wrapper
      .find('Button')
      .last()
      .simulate('click');
    expect(mockCallbackShutdown.mock.calls.length).toBe(1);
  });
});
