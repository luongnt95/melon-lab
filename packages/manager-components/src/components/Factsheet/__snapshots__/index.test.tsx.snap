// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Factsheet should render correctly 1`] = `
<div
  className="jsx-__scopedHash factsheet"
>
  <JSXStyle
    css="__scoped"
    styleId="__scopedHash"
  />
  <h3
    className="jsx-__scopedHash factsheet__title"
  >
    Melon Fund
    <a
      className="jsx-__scopedHash factsheet__title-link"
      href="https://twitter.com/intent/tweet?text=My%20%23MelonFund%20%22Melon%20Fund%22%20has%20a%20share%20price%20currently%20of%200.9824.%20Have%20a%20look%3A&url=https%3A%2F%2Fmelon.fund%2F%23undefined&hashtags=TechnologyRegulatedFunds%2CMelon%2CMelonFund&via=melonport&related=melonport"
      rel="noopener noreferrer"
      target="_blank"
    >
      <Icon
        height="12px"
        name="icons_twitter"
        width="12px"
      />
    </a>
  </h3>
  <div
    className="jsx-__scopedHash"
  >
    Creation date:
     
    <Loading
      loading={false}
    >
      24. Jul 2018 09:19
    </Loading>
    <br
      className="jsx-__scopedHash"
    />
    <Button
      onClick={[Function]}
      style="clear"
    >
      AUM:
       
      <Loading
        dataAvailable={false}
        loading={false}
      >
        0.9824
      </Loading>
       
      WETH-T
    </Button>
    <Button
      onClick={[Function]}
      style="clear"
    >
      Share price:
       
      <Loading
        dataAvailable={false}
        loading={false}
      >
        0.9824
      </Loading>
       
      WETH-T
      /Share
    </Button>
    <a
      className="jsx-__scopedHash"
      href="/"
    >
      Ranking: 
      <Loading
        loading={false}
      >
        1
      </Loading>
       out of
       
      <Loading
        loading={false}
      >
        12
      </Loading>
    </a>
    <br
      className="jsx-__scopedHash"
    />
    Total number of shares:
     
    <Loading
      loading={false}
    >
      1.0000
    </Loading>
    <br
      className="jsx-__scopedHash"
    />
    Shares owned by me:
     
    <Loading
      loading={false}
    >
      1
    </Loading>
    <hr
      className="jsx-__scopedHash"
    />
    Management Reward:
     
    <Loading
      loading={false}
    >
      0
    </Loading>
    %
    <br
      className="jsx-__scopedHash"
    />
    Performance Reward:
     
    <Loading
      loading={false}
    >
      0
    </Loading>
    %
    <hr
      className="jsx-__scopedHash"
    />
    <a
      className="jsx-__scopedHash"
      href="https://ipfs.io/ipfs/Qmc9JRw4zarrs6gJwu6tC58UAgeEujNg9VMWcH8MUEd5TW/"
      rel="noopener noreferrer"
      target="_blank"
    >
      Contact Investors/Managers
    </a>
    <hr
      className="jsx-__scopedHash"
    />
    <a
      className="jsx-__scopedHash"
      href="https://melon-reporting.now.sh/report/"
      rel="noopener noreferrer"
      target="_blank"
    >
      Generate fund report
    </a>
    <hr
      className="jsx-__scopedHash"
    />
    <Button
      onClick={[MockFunction]}
      style="clear"
    >
      Irreversibly shut down fund
    </Button>
  </div>
</div>
`;

exports[`Factsheet should render correctly with isCompetition 1`] = `
<div
  className="jsx-__scopedHash factsheet"
>
  <JSXStyle
    css="__scoped"
    styleId="__scopedHash"
  />
  <h3
    className="jsx-__scopedHash factsheet__title"
  >
    Melon Fund
    <a
      className="jsx-__scopedHash factsheet__title-link"
      href="https://twitter.com/intent/tweet?text=My%20%23MelonFund%20%22Melon%20Fund%22%20has%20a%20share%20price%20currently%20of%200.9824.%20Have%20a%20look%3A&url=https%3A%2F%2Fmelon.fund%2F%23undefined&hashtags=TechnologyRegulatedFunds%2CMelon%2CMelonFund&via=melonport&related=melonport"
      rel="noopener noreferrer"
      target="_blank"
    >
      <Icon
        height="12px"
        name="icons_twitter"
        width="12px"
      />
    </a>
  </h3>
  <div
    className="jsx-__scopedHash"
  >
    Creation date:
     
    <Loading
      loading={false}
    >
      24. Jul 2018 09:19
    </Loading>
    <br
      className="jsx-__scopedHash"
    />
    <Button
      onClick={[Function]}
      style="clear"
    >
      AUM:
       
      <Loading
        dataAvailable={false}
        loading={false}
      >
        0.9824
      </Loading>
       
      WETH-T
    </Button>
    <Button
      onClick={[Function]}
      style="clear"
    >
      Share price:
       
      <Loading
        dataAvailable={false}
        loading={false}
      >
        0.9824
      </Loading>
       
      WETH-T
      /Share
    </Button>
    <a
      className="jsx-__scopedHash"
      href="/"
    >
      Ranking: 
      <Loading
        loading={false}
      >
        1
      </Loading>
       out of
       
      <Loading
        loading={false}
      >
        12
      </Loading>
    </a>
    <br
      className="jsx-__scopedHash"
    />
    Total number of shares:
     
    <Loading
      loading={false}
    >
      1.0000
    </Loading>
    <br
      className="jsx-__scopedHash"
    />
    Shares owned by me:
     
    <Loading
      loading={false}
    >
      1
    </Loading>
    <hr
      className="jsx-__scopedHash"
    />
    Management Reward:
     
    <Loading
      loading={false}
    >
      0
    </Loading>
    %
    <br
      className="jsx-__scopedHash"
    />
    Performance Reward:
     
    <Loading
      loading={false}
    >
      0
    </Loading>
    %
    <hr
      className="jsx-__scopedHash"
    />
    <a
      className="jsx-__scopedHash"
      href="https://ipfs.io/ipfs/Qmc9JRw4zarrs6gJwu6tC58UAgeEujNg9VMWcH8MUEd5TW/"
      rel="noopener noreferrer"
      target="_blank"
    >
      Contact Investors/Managers
    </a>
    <hr
      className="jsx-__scopedHash"
    />
    <a
      className="jsx-__scopedHash"
      href="https://melon-reporting.now.sh/report/"
      rel="noopener noreferrer"
      target="_blank"
    >
      Generate fund report
    </a>
    <hr
      className="jsx-__scopedHash"
    />
  </div>
</div>
`;

exports[`Factsheet should render correctly with loading 1`] = `
<div
  className="jsx-__scopedHash factsheet"
>
  <JSXStyle
    css="__scoped"
    styleId="__scopedHash"
  />
  <h3
    className="jsx-__scopedHash factsheet__title"
  >
    Melon Fund
    <a
      className="jsx-__scopedHash factsheet__title-link"
      href="https://twitter.com/intent/tweet?text=My%20%23MelonFund%20%22Melon%20Fund%22%20has%20a%20share%20price%20currently%20of%200.9824.%20Have%20a%20look%3A&url=https%3A%2F%2Fmelon.fund%2F%23undefined&hashtags=TechnologyRegulatedFunds%2CMelon%2CMelonFund&via=melonport&related=melonport"
      rel="noopener noreferrer"
      target="_blank"
    >
      <Icon
        height="12px"
        name="icons_twitter"
        width="12px"
      />
    </a>
  </h3>
  <Spinner
    icon={true}
    size="small"
  />
</div>
`;
